src/tools/rbd_mirror/LeaderWatcher.cc

```diff
@@ -0,0 +1,712 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "LeaderWatcher.h"
+#include "common/Timer.h"
+#include "common/debug.h"
+#include "common/errno.h"
+#include "cls/rbd/cls_rbd_client.h"
+#include "librbd/Utils.h"
+#include "librbd/watcher/Types.h"
+#include "Threads.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_rbd_mirror
+#undef dout_prefix
+#define dout_prefix *_dout << "rbd::mirror::LeaderWatcher: " \
+                           << this << " " << __func__ << ": "
+
+namespace rbd {
+namespace mirror {
+
+using namespace leader_watcher;
+
+using librbd::util::create_context_callback;
+using librbd::util::create_rados_ack_callback;
+
+LeaderWatcher::LeaderWatcher(Threads *threads, librados::IoCtx &io_ctx)
+  : Watcher(io_ctx, threads->work_queue, RBD_MIRROR_LEADER),
+    m_threads(threads),
+    m_lock("rbd::mirror::LeaderWatcher " + io_ctx.get_pool_name()),
+    m_notifier_id(librados::Rados(io_ctx).get_instance_id()) {
+}
+
+int LeaderWatcher::init() {
+  dout(20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+  assert(!m_leader_lock);
+
+  int r = m_ioctx.create(m_oid, false);
+  if (r < 0) {
+    derr << "error creating " << m_oid << " object: " << cpp_strerror(r)
+         << dendl;
+    return r;
+  }
+
+  m_leader_lock.reset(
+    new LeaderLock(m_ioctx, m_work_queue, m_oid, this, true,
+                   g_ceph_context->_conf->rbd_blacklist_expire_seconds));
+
+  C_SaferCond register_ctx;
+  register_watch(&register_ctx);
+  r = register_ctx.wait();
+  if (r < 0) {
+    derr << "error registering leader watcher for " << m_oid << " object: "
+         << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  acquire_leader_lock();
+  return 0;
+}
+
+void LeaderWatcher::shut_down() {
```

Comment:
Might be better to make this an async method w/ helper methods now that you have to unlock/lock over and over around to get around the fact that it is internally calling async methods.